Sources/MongoSwift/BSON/Document.swift

```diff
@@ -239,20 +241,31 @@ public struct Document: ExpressibleByDictionaryLiteral, ExpressibleByArrayLitera
         get { return DocumentIterator(forDocument: self, advancedTo: key)?.currentValue }
         set(newValue) {
             do {
-                try self.setValue(for: key, to: newValue)
+                if let newValue = newValue {
+                    try self.setValue(for: key, to: newValue)
+                } else {
+                    // TODO SWIFT-224: use va_list variant of bson_copy_to_excluding to improve performance
+                    self = self.filter { $0.key != key }
+                }
             } catch {
                 preconditionFailure("Failed to set the value for key \(key) to \(newValue ?? "nil"): \(error)")
             }
         }
     }
 
+    /// An implementation identical to subscript(key: String), but offers the ability to choose a default value if the
+    /// key is missing.
+    public subscript(key: String, default defaultValue: @autoclosure () -> BSONValue) -> BSONValue {
```

Comment:
Is it common for this to use an autoclosure as the way to yield a default value? The tests seem to use basic types (e.g. a float value, newly constructed NSNull), so I'm a bit confused. Is this intended to support anything that would resolve to a BSONValue?

Reply:
@jmikola Technically, the only thing `@autoclosure` does is wrap arguments in closures. Since even single-value arguments are just expressions, we are technically using `@autoclosure`! By the way: even without `@autoclosure`, the function would _still_ support any expression that evaluated to a `BSONValue`, because proving that resolution is the compiler's job. This only makes it so that the actual evaluation happens later. As @kmahar points out, I only kept this because this is what `Dictionary` does.

I've added a more complicated case where we do `floatVal * floatVal` so that we have something that isn't just a trivial value. This isn't much more complex, but is at the very least an expression that will be have its time of computation affected. Let me know if you guys want something even crazier.

Reply:
tests look sufficiently complex to me.